licensing/src/main/java/io/crate/license/DecryptedLicenseData.java

```diff
@@ -53,23 +64,68 @@ String issuedTo() {
         return issuedTo;
     }
 
+    String signature() {
+        return (signature.isPresent() ? signature.get() : "");
+    }
+
+    void setSignature(String signature) {
+        this.signature = Optional.ofNullable(signature);
+    }
+
+    boolean isExpired() {
+        return expirationDateInMs < System.currentTimeMillis();
+    }
+
     /*
      * Creates the json representation of the license information with the following structure:
      *
      * <pre>
      *      {
      *          "expirationDateInMs": "XXX",
-     *          "issuedTo": "YYY"
+     *          "issuedTo": "YYY",
+     *          "signature": "ZZZ"
+     *      }
+     * </pre>
+     *
+     * If no signature is set (eg. for Self Generated Licenses), the signature part is omitted e.g.
+     * <pre>
+     *      {
+     *          "expirationDateInMs": "XXX",
+     *          "issuedTo": "YYY",
      *      }
      * </pre>
      */
     byte[] formatLicenseData() {
+        // by default include signature if already set
+        return formatLicenseData(signature.isPresent());
+    }
+
+    /*
+     * Creates the json representation of the license information for signature generation.
+     * This is always have the following structure:
+     *
+     * <pre>
+     *      {
+     *          "expirationDateInMs": "XXX",
+     *          "issuedTo": "YYY",
+     *      }
+     * </pre>
+     */
+    byte[] formatLicenseDataForSignature() {
+        // never include signature
+        return formatLicenseData(false);
+    }
+
+    private byte[] formatLicenseData(boolean includeSignature) {
         try {
             XContentBuilder contentBuilder = XContentFactory.contentBuilder(XContentType.JSON);
             contentBuilder.startObject()
                 .field(EXPIRATION_DATE_IN_MS, expirationDateInMs)
-                .field(ISSUED_TO, issuedTo)
-                .endObject();
+                .field(ISSUED_TO, issuedTo);
+            if (includeSignature) {
```

Comment:
Shouldn't we always include it if it is present? Why is it okay to throw away the signature?

Reply:
At the time of the verification, where the signature is present at the license, we need to generate a new signature based on the  license content (exluding the signature itself) so that we test signature validity. In that case we need to skip the signature.

It was the reason that I insisted of having another class i.e. LicenseInfo containing only the actual license information (ie expirationDate, issuedTo). I think this wouls be much cleaner that way.